File: 2022/11/11.py
import sys, math, operator

monkeys = []

class Monkey:
    modulo = 1
    def __init__(self, monkey: str):
        _, items, op, test, if_t, if_f = monkey.split("\n")
        self.items = [int(a) for a in items.split(":")[1].split(',')]
        ops = {"*": operator.mul, "+": operator.add}
        *_, op, arg = op.split(":")[1].split()
        self.op = lambda x: ops[op](x, x if arg == "old" else int(arg))
        self.divisible_by = int(test.split()[-1])
        Monkey.modulo *= self.divisible_by
        self.if_t = int(if_t.split()[-1])
        self.if_f = int(if_f.split()[-1])
        self.passes = 0

    def round(self, div_by_3: bool):
        for item in self.items:
            new_worry = self.op(item) // (3 if div_by_3 else 1) % Monkey.modulo
            self.passes += 1
            new_index = self.if_t if new_worry % self.divisible_by == 0 else self.if_f
            monkeys[new_index].items.append(new_worry)
        self.items.clear()


monkeys2 = []
for i, monkey in enumerate(sys.stdin.read().strip().split("\n\n")):
    monkeys.append(Monkey(monkey))
    monkeys2.append(Monkey(monkey))

for i in range(20):
    for monkey in monkeys:
        monkey.round(True)
print(math.prod(sorted([a.passes for a in monkeys])[-2:]))

monkeys = monkeys2
for i in range(10000):
    for monkey in monkeys:
        monkey.round(False)
print(math.prod(sorted([a.passes for a in monkeys])[-2:]))

File: 2022/11/11.kt
class Monkey(monkey: String) {
    var counter = 0L
    var items: MutableList<Long>
    var op: (Long) -> Long
    val divisibleBy: Long
    val idIfTrue: Int
    val idIfFalse: Int

    companion object {
        var modulo: Long = 1
        var monkeys: MutableList<Monkey> = mutableListOf()
    }

    init {
        val attributes = monkey.split("\n").map { it.split(":")[1].trim() }
        val (_, _, _, opMultOrAdd, opArg) = attributes[2].split(" ")
        val opFunc: (Long, Long) -> Long = if (opMultOrAdd == "*") Long::times else Long::plus

        items = attributes[1].split(",").map { it.trim().toLong() }.toMutableList()
        op = { x -> opFunc(x, opArg.toLongOrNull() ?: x) }
        divisibleBy = attributes[3].split(" ").last().toLong()
        idIfTrue = attributes[4].split(" ").last().toInt()
        idIfFalse = attributes[5].split(" ").last().toInt()
        Monkey.modulo *= divisibleBy
        Monkey.monkeys.add(this)
    }
    fun round(divideBy: Int = 1) {
        for (item in items) {
            val newWorry = op(item) % Monkey.modulo / divideBy
            val index = if (newWorry % divisibleBy == 0L) idIfTrue else idIfFalse
            Monkey.monkeys[index].items.add(newWorry)
            counter++
        }
        items.clear()
    }
}

fun main() {
    val monkeys = generateSequence(::readlnOrNull).joinToString("\n").trim().split("\n\n")

    fun multLast2() = Monkey.monkeys.map { it.counter }.sorted().takeLast(2).reduce(Long::times)
    fun rounds(count: Int, divBy: Int) = repeat(count) { Monkey.monkeys.map { it.round(divBy) } }

    // Part 1
    monkeys.map { Monkey(it) }
    rounds(20, divBy=3)
    println(multLast2())

    // Part 2
    Monkey.modulo = 1
    Monkey.monkeys.clear()
    monkeys.map { Monkey(it) }
    rounds(10000, divBy=1)
    println(multLast2())
}
